Draft for Spring Boot 3 WAR post (#3655)
OpenLiberty/blogs: posts/2024-xx-xx-spring-boot-3.adoc

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/scottkurz
author_github: https://github.com/scottkurz
seo-title: Running a Spring Boot 3.x application WAR file on Liberty
seo-description: TODO In version 23.0.0.9 and later, you can easily run a Spring Boot 3.x application from a WAR file on Open Liberty. This post use the Spring Boot Petclinic sample application and the Liberty Maven plugin to create and run a WAR-based Spring Boot 3.1 application on Open Liberty. 
blog_description: TODO In version 23.0.0.9 and later, you can easily run a Spring Boot 3.x application from a WAR file on Open Liberty. This post uses the Spring Boot Petclinic sample application and the Liberty Maven plugin to create and run a WAR-based Spring Boot 3.1 application on Open Liberty. 
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= Running a Spring Boot 3.x application WAR file on Liberty
Scott Kurz <https://github.com/scottkurz>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> As of Liberty version 23.0.0.9 there are now two ways to deploy Spring Boot WAR files on Liberty.

Reviewer: I know this hasn't been addressed in any of our other blog posts/guide about SB support yet either, but I think this sentence should be followed by answering the question of "why would I do that [deploy a Spring Boot WAR on Liberty]?". Without it, yes, the post is providing some helpful documentation for people who need it, but a blog post should provide more than that and it would be good to also convince readers that it's a valuable thing to do for relevant use cases.

Reviewer: eg to benefit from InstantOn for a faster startup whilst maintaining throughput and not having to modify the app?

Author: The answer to "why" might depend on your baseline alternative

 * not going to get into why choose WAR but assume they know.  I even feel like "..to take advantage of the Spring and Spring Boot libraries " is too obvious to mention but leaving it for now as a compromise
 * follow-up.. this good to reference: 

That answers why SB on Liberty.

As far as why "optimized deployment"?  That's addressed with:

>  This deployment adds some additional configuration options, and provides advantages for efficiently building Docker container images

OTOH, InstantOn is supposed to work with both deployments.

> While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, it can also be deployed using an optimized deployment for Spring Boot applications. This deployment adds some additional configuration options, and provides advantages for efficiently building Docker container images.

Reviewer: ```suggestion
While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, the WAR can be optimized deployment for Spring Boot applications. The optimization adds some additional configuration options, and provides advantages for efficiently building Docker container images.
```

Reviewer: Also, why would you ever use a non-optimized approach? It sounds, um, sub-optimal...

Author: * This part "..the WAR can be optimized deployment .." read a bit awkward so I took your suggestion and tweaked a bit

"Also, why would you ever use a non-optimized approach? It sounds, um, sub-optimal..."

Because:
 * you need extra pom.xml config
 * you need special server.xml config
 * you can't use dev mode

and because you might want to deploy like any WAR.

I added:
"using some extra pom.xml and server.xml configuration."

to maybe thread the needle here or feel free to suggest another wording.

> In this blog post we will step through a simple flow starting from the Spring Initializr.  We will deploy this to Liberty via a standard WAR deployment without any Liberty modifications to the app.  

Reviewer: ```suggestion
In this blog post, we create a simple app by using the Spring Initializr, and then deploy the app as a WAR to Liberty without any modifications.
```

Author: Trying to emphasize the "typical deployment" contrasted with the later optimized, so added to your wording"

"..., using the same deployment as for any WAR"

> From there we'll show how to deploy using the optimized Spring Boot deployment by adding the liberty-maven-plugin and some minimal config to the pom.xml, and simple server.xml configuration using the `springBoot-3.0` feature.   We will then be able to build an efficiently-layered Docker image.

Reviewer: ```suggestion
We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and some minimal config to the `pom.xml`, and add a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
```

Author: Mostly took your change ..but left the words "the optimized deployment".  This deployment is a direction that drags in several ramifications.  When you go this route, you have to have special pom.xml, server.xml config.. 
It even affects a bit of the programming model.

It's not just something you do in the final step of deployment.  It's a choice you have to make.

> Much of the Liberty documentation relating to Spring Boot focuses on the optimized deployment associated with the Spring Boot features, and so we spend some extra time in this post explaining both options.

Reviewer: Can you add a link to that Liberty doc?

Reviewer: And also say why the Liberty doc focuses mainly on the optimized approach and, conversely, why it's worth covering both in this post? Otherwise it sounds like you're criticising the Liberty doc (instead of just fixing the doc), and it's not clear why the Liberty doc would do one thing and you'd do something else in the blog post?

Author: Done.   Also opened https://github.com/OpenLiberty/docs/issues/7296 for doc follow-up... and had opened https://github.com/OpenLiberty/docs/issues/7284 for tracking the major.minor version support. (edited)

== Prerequisites
- Java 21 (sample could be easily reworked with Java 17)

> == Deploy as standard WAR

Reviewer: I know this is an initial draft, but just so we don't forget: expand the sentences into full sentences throughout.

Author: Done

+
2. Add web controller `vi src/main/java/demo/HelloController.java`
+
[source,java]
----
package demo;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class HelloController {

    @GetMapping("/")
    public String hello() {
        return "Hello from Spring Boot in Open Liberty!";
    }

}
----
+
3. Run with standard WAR deployment with default configuration
+
[source,sh]
----
./mvnw io.openliberty.tools:liberty-maven-plugin:3.10:dev -Dtemplate=webProfile10 -DskipTests
----
+
4. Test the application by visiting the following URL in a browser: http://localhost:9080/spring-liberty-0.0.1-SNAPSHOT/
+
5. Stop the server (with  `<Ctrl>+C` or `q,<Enter>` to quit dev mode).
+
6. Recap
+
The app is bootstrapped by the web container via the ServletInitializer, extending `SpringBootServletInitializer`.  We go out of our way to run this example with only default Liberty and Maven configuration even though it'd be more typical to maintain a server.xml configuration for your app.

> == Use Liberty Spring Boot feature deployment

Reviewer: I don't think I understand what this heading means and how it relates back to what you discussed in the intro. Does this map to the "optimized deployment" approach?

Author: Reworded.  Yes, exactly.  I was hoping the reader would make the leap to equate the two ("optimized" == "Spring Boot feature") but the fact that you asked shows that was wrong...should be more explicit

We're going to reuse the app from the previous example.

> 1. Add liberty-maven-plugin with Spring Boot deployment config to pom.xml:

Reviewer: ok, but what does this do for the app?

+
[source,xml]
----
    <plugin>
        <groupId>io.openliberty.tools</groupId>
        <artifactId>liberty-maven-plugin</artifactId>
        <version>3.10</version>
        <configuration>
            <appsDirectory>apps</appsDirectory>
            <deployPackages>spring-boot-project</deployPackages>
        </configuration>
    </plugin>
----
+

> 2. Create your Liberty server.xml by copying from the Spring Boot template

Reviewer: And what does this do for the app?

Author: Did my extra sentence address your concern?

+
[source,sh]
----
./mvnw clean liberty:create -Dtemplate=springBoot3
mkdir -p src/main/liberty/config/
cp target/liberty/wlp/usr/servers/defaultServer/server.xml src/main/liberty/config/server.xml
----
+

> 3. Add application arguments to your Spring Boot application configuration in Liberty server.xml

Reviewer: What do these do for the app?

Author: Did my extra sentence address your concern?

+
[source,xml]
----
 <springBootApplication id="spring-liberty" location="thin-spring-liberty-0.0.1-SNAPSHOT.war" name="spring-liberty">
    <applicationArgument>--server.servlet.context-path=/testpath</applicationArgument>
    <applicationArgument>--myarg=myval</applicationArgument>
 </springBootApplication>
----
+
4. Add println to echo args in output
+
[source,java]
----
	public static void main(String[] args) {
        System.out.println("Arguments array = " + java.util.Arrays.toString(args));
		SpringApplication.run(SpringLibertyApplication.class, args);
	}
----
+
5. Run
+

> Though we won't benefit from using dev mode, we can still use the "lifecycle" built into the 'run' goal.  We do need to do a 'package' first to get the Spring Boot "repackage" packaging.

Reviewer: Why not? (I know dev mode doesn't support SB yet but I think that needs to be said explicitly (somewhere) otherwise it's just a bit mysterious.)

Author: Explained

+
[source,sh]
----
./mvnw package liberty:run
----
+

> 6.  Observe the output:

Reviewer: What if the reader/user doesn't see that output? What's most likely gone wrong?

Author: Good question.  I don't have an answer though.  If I think of one I'll add it

+
[source,sh]
----
[INFO] Arguments array = [--server.servlet.context-path=/testpath, --myarg=myval]
[INFO]   .   ____          _            __ _ _
[INFO]  /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
[INFO] ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
[INFO]  \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
[INFO]   '  |____| .__|_| |_|_| |_\__, | / / / /
[INFO]  =========|_|==============|___/=/_/_/_/
[INFO]  :: Spring Boot ::                (v3.2.4)
----
+
and also
+
[source,sh]
----
[INFO] [AUDIT   ] CWWKT0016I: Web application available (default_host): http://localhost:9080/testpath/
----
+

> 7. Test the application by visiting the following URL in a browser: http://localhost:9080/testpath/

Reviewer: What should you see at that URL?

Author: TODO - I owe you an answer.  Will set up and run

Author: DONE

+
8. When you finish testing the application, stop the server by typing `<Ctrl>+C`.
+
9. Recap
+
The server is configured with the https://openliberty.io/docs/latest/reference/feature/springBoot-3.0.html[springBoot-3.0] feature. The liberty-maven-plugin has been added to the pom.xml along with special configuration to use the optimized Spring Boot deployment.  The app is packaged as an "executable WAR" by running the `spring-boot:repackage` goal in the `package` phase, and the app is bootstrapped via its `main()` method in `SpringLibertyApplication` passing in application arguments defined in `server.xml`.
+

> If you're wondering, yes, the "executable WAR" binary package here could still be deployed as a standard WAR (though a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment).

Reviewer: ```suggestion
If you're wondering, yes, while a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment, that executable WAR could still be deployed as a standard WAR.
```

Author: Took

> == Build container image with efficient layering

Reviewer: Should this be an H3, rather than H2? It seems to relate only to the optimized instructions?

Author: Good catch, you're thorough :)
